tests/modules/users/test_permissions.py

```diff
@@ -443,6 +444,24 @@ def test_ObjectAccessPermission_contributor_user(
     validate_cannot_write_object(owned_encounter)
     validate_cannot_delete_object(owned_encounter)
 
+    # project, encounter, asset permitted via project
+    project1 = Mock(is_public=lambda: False, get_encounters=lambda: [])
+    project2 = Mock(is_public=lambda: False)
+    project2.get_encounters.return_value = [Mock(get_assets=lambda: []), owned_encounter]
+    mock_asset = Mock(__class__=Asset, is_public=lambda: False)
+    project3 = Mock(is_public=lambda: False)
+    project3.get_encounters.return_value = [Mock(get_assets=lambda: [mock_asset])]
+    # object that is not a project, encounter or asset
+    mock_other = Mock(is_public=lambda: False)
+    projects = [project1, project2, project3]
+    with patch.object(contributor_1_login, 'owns_object', return_value=False):
```

Comment:
I really like all this, it looks extremely powerful. 
One question though, is this patching of contributor_1_login persistent?
In other words, once this test is run, does it change how contributor_1_login would behave for all subsequent tests?
Therefore does the test output depend on the order that tests are run in?

Reply:
@hwindsor That's a very good question!  I patched `contributor_1_login.owns_object` using a context manager / `with` statement, which means the patch only applies until the block exits.  So it doesn't apply after line 464.